### PR TITLE
Provide tab completion using Click

### PR DESCRIPTION
**Pull request by MaxG87**
This merge provides tab completion for Bash and Zsh using the capabilities of the click command line parsing framework. This fixes the issues #269, #270 and supersedes the pull requests #267 and #268.

Currently the new approach lacks many of the capabilities of the manually created completion scripts. However, given the already added completions for tags and projects, it should be easy to improve the generated completions of the click framework.

### CODE REVIEW

**Comment by MaxG87** [comment]
Please consider this PR more as a request for comment than as a real pull request. I managed to get the completions working, but I would like to know what you think of the code and the workflow before I dive deeper. Especially, I would like to know where you want the tab completion helpers, currently get_projects and get_tags, to live.

**Comment by k4nar** [comment]
This looks great! Does it provide good performance? And do you have a list of what's missing to be of par with the current completion?

**Comment by MaxG87** [comment]
The performance leaves nothing to be desired for me. Missing features are:

  * add: completion of projects and tags
  * edit: completion of frames
  * help: completion of commands
  * remove: completion of frames 
  * rename: completion of type, projects and tags
  * restart: completion of frames
  * start: completion of projects and tags

Some of the condidional compeltions seem to be the hardest to achive. For
example, `watson start` has a quite complicated logic to determine whats a
project and whats a tag. I do not yet know how to autocomplete this.

However, all options/flags starting with a dash are autocompleted now. This is
already an improvement to before.

**Comment by jmaupetit** [comment]
> Some of the condidional compeltions seem to be the hardest to achive. For
> example, watson start has a quite complicated logic to determine whats a
> project and whats a tag. I do not yet know how to autocomplete this.

That's unfortunate. IMO this is a requirement. We need some help here!

**Comment by MaxG87** [comment]
I wrote some of the easy autocompletions. Now the missing pieces are:

  * add: completion of projects and tags
  * help: completion of commands
  * rename: projects and tags
  * start: completion of projects and tags

I currently use Watson with my click completion, thus without completion for projects and tags in `watson start`. I definitely can second that there tab completion is a must.

I am certain that I can complete tab completion for `watson rename` soon. For `watson start` and `watson add` I am certain that it will work _somehow_ but I have no clue yet. So it will take some time.

Again, please comment on the code if you like too, so I can already create it in the right style.

**Comment by MaxG87** [comment]
I wrote the hard part. Now there is tab completion for projects and tags in `watson start` and `watson add`. Only command completion for `watson help` is missing, which I would hope to be not supercritical.

What do you think of it?

**Comment by MaxG87** [comment]
I wanted to ask what the status of this PR is. Are there any improvements to be done by me? I am using this feature for some weeks now and I did not found too many rough edges.

**Comment by MaxG87** [comment]
I am very sorry for not working on this for so long. But now I can finish it, hopefully.

> I think the code should be moved to a separate module (.autocomplete ?) for clarity.
Done.

> Also, do you think we could add some unit tests to the main functions?
I could need some assistance here. All functions require an instance of watson.Watson() to be available. Do you have a mock for this already? If not, should I include a JSON to tests/resources and tweak watson.Watson until I can initialise it using that JSON?

**Comment by k4nar** [comment]
> I could need some assistance here. 

No problem!

> All functions require an instance of watson.Watson() to be available. Do you have a mock for this already? If not, should I include a JSON to tests/resources and tweak watson.Watson until I can initialise it using that JSON?

There is a pytest fixture for it [here](https://github.com/TailorDev/Watson/blob/master/tests/conftest.py#L13-L15). You can use it [like this](https://github.com/TailorDev/Watson/blob/master/tests/test_watson.py#L221-L228) for instance. Everything will be removed between each test.

**Comment by MaxG87** [comment]
Writing tests is much harder than I thought. After setting up the fixture I need, I noticed that the code to test initialises a new Watson instance based on 'os.environ'. I managed to use the monkeypatch fixture to change that.

Unfortunately, I do not now how to instruct pytest-datafiles (the `pytest.mark.datafiles`) to move a file to the temporary directory and rename it there. In particular, I would like to use `tests/resources/frames-for-autocompletion` for the unit tests. However, I only manage to move it to the test config-dir *without* changing its name.

Could you please help me out here? How could I achieve this?

**Comment by MaxG87** [comment]
I managed to work around the problem with the file renamings. The test suite should be fairly comprehensive, but I would be glad to learn about suggestions for improvements.

I hope this can be merged now, I want to switch back to using the mainline Watson :-).

**Comment by jmaupetit** [comment]
@MaxG87 can you rebase your branch with the master branch (and resolve conflicts) so that we can merge your work? :pray:

**Comment by MaxG87** [comment]
Upon a kind hint by @davidag I have adapted the procedure on how to deploy completion receipts. It now is in line with the documentation again.

However, the make receipt now requires that watson can be found in PATH. I find that acceptable and not worth the complexity of further sanity checks. Please let me know if you think otherwise.

**Comment by Asdolo** [comment]
Hi! What's the status of this?

**Comment by MaxG87** [comment]
Imho, it is ready to be merged. I use it productively.

File: docs/index.md [hunk 1]
```diff
@@ -84,8 +84,18 @@ $ python setup.py install
...
 ### Command line completion
 
+#### Generating Completion Receipts
+
+The completion receipts for both Bash and ZSH can be generated by
+
+    make completion-scripts
+
+However, this receipt requires to have `watson` available in your PATH
+environment variable.
```
Comment (jmaupetit) [inline docs/index.md]: I think auto-completion files should be distributed with Watson. One should update them while packaging a new release. Can you also update the contributing documentation [1] to add this step in the release process?

[1] http://tailordev.github.io/Watson/contributing/release-new-version/

**Comment by MaxG87** [comment]
I just resolved the recent conflicts. I would be very glad if this could be merged (or rejected) soon, because it is quite annoying to do so regularly. Also, I would like to start to work on other contributions.

**Pull request merged**